File: append-etl/etl-append-data-to-tables.scala
// start your spark shell

// This script is designed to get your team started working with data append
// You may set the paths and paste it into spark shell via paste mode, :paste followed by ctrl-D to run
// Or bake it into a script and run it via spark-submit
// The script reads in the data append and taxonomy, processes the data append, and saves the results as parquet files
// The schemas are essentially ID, attribute, and values are the selected value for the attribute
// Single selects are things like "which one of the the following options" - so the values are the selected option
// Single selects may still have null values if we assess the question does not apply to the record
// Multi selects are "yes or no" conditions, so the values are 0 or 1

import org.apache.spark.sql._
import org.apache.spark.sql.functions._

// Define parameters for data paths and output directory
case class Params(
                   dataAppendPath: String = "path to your dataAppend file (csv.gz)",
                   dataAppendTaxonomyPath: String = "path to your data append taxonomy (csv)",
                   outDir: String = "where you want to save your files, as parquet"
                 )

val params = Params()

// Define UDFs

// UDF to split the 'ATTRIBUTES' column into an array of svkeys
val splitSvkeys = udf((attributes: String) => attributes.split(",\\s*"))


// Read in the data assets
val dataAppend = spark.read
  .option("header", "true")
  .option("inferSchema", "true")
  .csv(params.dataAppendPath)

val dataAppendTaxonomy = spark.read
  .option("header", "true")
  .option("inferSchema", "true")
  .csv(params.dataAppendTaxonomyPath)

// Process 'dataAppend' by splitting 'ATTRIBUTES' into 'trueSvkeys' and cache the result
val dataAppendWithSvkeys = dataAppend
  .withColumn("trueSvkeys", splitSvkeys(col("ATTRIBUTES")))
  .cache()

// Trigger an action to cache the DataFrame
dataAppendWithSvkeys.count()

// Narrow the taxonomic scope by extracting distinct svkeys
val superSetSvkeys = dataAppendWithSvkeys
  .select(explode(col("trueSvkeys")).as("svkey"))
  .distinct()
  .as[String]
  .collect()

// Filter taxonomy based on the relevant svkeys and create 'attributeString'
val focalTaxonomy = dataAppendTaxonomy
  .where(col("survey_value_key").isin(superSetSvkeys: _*))
  .withColumn("attributeString", concat_ws("_", col("AttributeKey"), col("survey_value_key")))

// Create a map for single select attributes
val singleSelectColumnMap: Map[String, String] = focalTaxonomy
  .filter(col("Attribute Type") === "Single Select")
  .select("survey_value_key", "AttributeKey")
  .distinct()
  .as[(Int, String)]
  .collect()
  .map { case (svkey, attrKey) => svkey.toString -> attrKey.toString }
  .toMap


// Extract distinct single and multi-select attributes
val singleSelectAttributes = singleSelectColumnMap.values.toArray.distinct
val multiSelectAttributes = focalTaxonomy
  .filter(col("Attribute Type") === "Multi Select")
  .select("attributeString")
  .distinct()
  .as[String]
  .collect()

val allAttributes = focalTaxonomy
  .select("attributeString")
  .distinct()
  .as[String]
  .collect()

// Broadcast the singleSelectColumnMap for efficient lookup in UDF
val broadcastMap = spark.sparkContext.broadcast(singleSelectColumnMap)
// UDF to map svkeys to a single select attribute
val mapSvkeyToAttr = udf((trueSvkeys: Seq[String], attr: String) => {
  // Access the broadcasted map within the UDF's closure
  trueSvkeys
    .flatMap(svkey => broadcastMap.value.get(svkey))
    .find(_ == attr)
    .getOrElse(null)
})

// Create DataFrame for single selects by adding a new column for each single select attribute
val singleSelectDF = singleSelectAttributes.foldLeft(dataAppendWithSvkeys) { (df, attr) =>
  df.withColumn(
    attr,
    mapSvkeyToAttr(col("trueSvkeys"), lit(attr))
  )
}.drop("trueSvkeys", "ATTRIBUTES")

// Process multi-select attributes
val multiSelectDF = dataAppendWithSvkeys
  .select(col("ID"), explode(col("trueSvkeys")).as("survey_value_key"))
  .join(focalTaxonomy, "survey_value_key")
  .groupBy("ID")
  .pivot("attributeString", multiSelectAttributes)
  .agg(count("survey_value_key"))
  .na.fill(0)

// Alternatively, process all attributes as multi-selects
val allAsMultiSelectDF = dataAppendWithSvkeys
  .select(col("ID"), explode(col("trueSvkeys")).as("survey_value_key"))
  .join(focalTaxonomy, "survey_value_key")
  .groupBy("ID")
  .pivot("attributeString", allAttributes)
  .agg(count("survey_value_key"))
  .na.fill(0)

// Save the processed DataFrames
singleSelectDF.write.mode("overwrite").save(params.outDir + "singleSelectDF")
multiSelectDF.write.mode("overwrite").save(params.outDir + "multiSelectDF")
allAsMultiSelectDF.write.mode("overwrite").save(params.outDir + "allAsMultiSelectDF")

File: append-etl/etl-append-data-to-tables.py
import pandas as pd
import numpy as np
import os
import argparse
import sys
import logging

def split_svkeys(attributes):
    """Split the 'ATTRIBUTES' column into a list of svkeys."""
    if pd.isna(attributes):
        return []
    return [svkey.strip() for svkey in attributes.split(",")]

def map_svkey_to_attr(true_svkeys, attr, single_select_map):
    """
    Map svkeys to a single select attribute.
    Returns the first matching svkey or None.
    """
    for svkey in true_svkeys:
        if svkey in single_select_map and single_select_map[svkey] == attr:
            return svkey  # or another desired value from the mapping
    return None

def parse_arguments():
    """Parse command-line arguments."""
    parser = argparse.ArgumentParser(description="ETL script to process dataAppend and taxonomy files and save as Parquet.")

    parser.add_argument(
        "--data_append_path",
        type=str,
        required=True,
        help="Path to your dataAppend file (csv.gz). Supports S3 URIs."
    )

    parser.add_argument(
        "--data_append_taxonomy_path",
        type=str,
        required=True,
        help="Path to your data append taxonomy file (csv). Supports S3 URIs."
    )

    parser.add_argument(
        "--out_dir",
        type=str,
        required=True,
        help="Directory where you want to save your output Parquet files. Supports S3 URIs."
    )

    return parser.parse_args()

def main():
    # Configure logging
    logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

    args = parse_arguments()

    data_append_path = args.data_append_path
    data_append_taxonomy_path = args.data_append_taxonomy_path
    out_dir = args.out_dir

    # Ensure output directory exists (for S3, this step is not strictly necessary)
    # os.makedirs(out_dir, exist_ok=True)  # Commented out since S3 handles directories implicitly

    # Read in the data assets
    logging.info("Reading dataAppend...")
    try:
        data_append = pd.read_csv(data_append_path, compression='gzip', header=0, dtype=str)
    except Exception as e:
        logging.error(f"Failed to read dataAppend: {e}")
        sys.exit(1)

    logging.info("Reading dataAppendTaxonomy...")
    try:
        data_append_taxonomy = pd.read_csv(data_append_taxonomy_path, header=0, dtype=str)
    except Exception as e:
        logging.error(f"Failed to read dataAppendTaxonomy: {e}")
        sys.exit(1)

    # Process 'dataAppend' by splitting 'ATTRIBUTES' into 'trueSvkeys'
    logging.info("Processing 'dataAppend' by splitting 'ATTRIBUTES'...")
    data_append['trueSvkeys'] = data_append['ATTRIBUTES'].apply(split_svkeys)

    # Extract distinct svkeys
    logging.info("Extracting distinct svkeys...")
    super_set_svkeys = pd.Series([svkey for sublist in data_append['trueSvkeys'] for svkey in sublist]).unique().tolist()

    # Filter taxonomy based on the relevant svkeys and create 'attributeString'
    logging.info("Filtering taxonomy and creating 'attributeString'...")
    focal_taxonomy = data_append_taxonomy[data_append_taxonomy['survey_value_key'].isin(super_set_svkeys)].copy()
    focal_taxonomy['attributeString'] = focal_taxonomy['AttributeKey'] + "_" + focal_taxonomy['survey_value_key'].astype(str)

    # Create a map for single select attributes
    logging.info("Creating singleSelectColumnMap...")
    single_select = focal_taxonomy[focal_taxonomy['Attribute Type'] == "Single Select"]
    single_select_column_map = pd.Series(single_select.AttributeKey.values, index=single_select.survey_value_key.astype(str)).to_dict()

    # Extract distinct single and multi-select attributes
    logging.info("Extracting distinct single and multi-select attributes...")
    single_select_attributes = list(single_select_column_map.values())

    multi_select = focal_taxonomy[focal_taxonomy['Attribute Type'] == "Multi Select"]
    multi_select_attributes = multi_select['attributeString'].unique().tolist()

    all_attributes = focal_taxonomy['attributeString'].unique().tolist()

    # Create singleSelectDF by adding a new column for each single select attribute
    logging.info("Creating singleSelectDF...")
    for attr in single_select_attributes:
        data_append[attr] = data_append['trueSvkeys'].apply(lambda svkeys: map_svkey_to_attr(svkeys, attr, single_select_column_map))

    single_select_df = data_append.drop(columns=['trueSvkeys', 'ATTRIBUTES'])

    # Save singleSelectDF as Parquet
    logging.info("Saving singleSelectDF...")
    try:
        single_select_df.to_parquet(os.path.join(out_dir, "singleSelectDF.parquet"), index=False)
    except Exception as e:
        logging.error(f"Failed to save singleSelectDF: {e}")
        sys.exit(1)

    # Process multi-select attributes
    logging.info("Processing multi-select attributes...")
    # Explode the 'trueSvkeys'
    exploded = data_append.explode('trueSvkeys')
    exploded = exploded.rename(columns={'trueSvkeys': 'survey_value_key'})

    # Merge with focal_taxonomy on 'survey_value_key'
    merged = exploded.merge(focal_taxonomy, on='survey_value_key', how='left')

    # Create pivot table for multi-select attributes
    logging.info("Creating multiSelectDF...")
    multi_select_df = merged.pivot_table(
        index='ID',
        columns='attributeString',
        aggfunc='size',
        fill_value=0
    ).reset_index()

    # Ensure all multi_select_attributes are present
    for attr in multi_select_attributes:
        if attr not in multi_select_df.columns:
            multi_select_df[attr] = 0

    multi_select_df = multi_select_df[['ID'] + multi_select_attributes]

    # Save multiSelectDF as Parquet
    logging.info("Saving multiSelectDF...")
    try:
        multi_select_df.to_parquet(os.path.join(out_dir, "multiSelectDF.parquet"), index=False)
    except Exception as e:
        logging.error(f"Failed to save multiSelectDF: {e}")
        sys.exit(1)

    # Alternatively, process all attributes as multi-selects
    logging.info("Processing all attributes as multi-selects...")
    all_as_multi_select_df = merged.pivot_table(
        index='ID',
        columns='attributeString',
        aggfunc='size',
        fill_value=0
    ).reset_index()

    # Ensure all_attributes are present
    for attr in all_attributes:
        if attr not in all_as_multi_select_df.columns:
            all_as_multi_select_df[attr] = 0

    all_as_multi_select_df = all_as_multi_select_df[['ID'] + all_attributes]

    # Save allAsMultiSelectDF as Parquet
    logging.info("Saving allAsMultiSelectDF...")
    try:
        all_as_multi_select_df.to_parquet(os.path.join(out_dir, "allAsMultiSelectDF.parquet"), index=False)
    except Exception as e:
        logging.error(f"Failed to save allAsMultiSelectDF: {e}")
        sys.exit(1)

    logging.info("ETL process completed successfully.")

if __name__ == "__main__":
    main()
